File: typescript/tests/debug/session-archive-v2.test.ts
import { describe, it, expect } from 'vitest';
import { gunzipSync, gzipSync } from 'node:zlib';
import { DebugSessionRegistry } from '../../src/debug/debug-session-registry.js';
import { SessionArchiveWriter } from '../../src/debug/archive/session-archive-writer.js';
import { SessionArchiveReader } from '../../src/debug/archive/session-archive-reader.js';
import {
  CURRENT_VERSION,
  MIN_SUPPORTED_VERSION,
  type SessionArchiveV2,
} from '../../src/debug/archive/session-archive.js';
import { computeMetadata } from '../../src/debug/archive/session-metadata.js';
import { PetriNet } from '../../src/core/petri-net.js';
import { Transition } from '../../src/core/transition.js';
import { place } from '../../src/core/place.js';
import { one } from '../../src/core/in.js';
import { outPlace } from '../../src/core/out.js';
import type { NetEvent } from '../../src/event/net-event.js';

const INPUT = place<string>('Input');
const OUTPUT = place<string>('Output');

const TEST_NET = PetriNet.builder('ArchiveV2TestNet')
  .transition(
    Transition.builder('Process')
      .inputs(one(INPUT))
      .outputs(outPlace(OUTPUT))
      .build()
  )
  .build();

function transitionEnabled(name: string, ts: number): NetEvent {
  return { type: 'transition-enabled', timestamp: ts, transitionName: name };
}

function transitionStarted(name: string, ts: number): NetEvent {
  return {
    type: 'transition-started',
    timestamp: ts,
    transitionName: name,
    consumedTokens: [{ value: 'payload', createdAt: ts }],
  };
}

function transitionCompleted(name: string, ts: number): NetEvent {
  return {
    type: 'transition-completed',
    timestamp: ts,
    transitionName: name,
    producedTokens: [{ value: 'result', createdAt: ts }],
    durationMs: 20,
  };
}

function transitionFailed(name: string, ts: number): NetEvent {
  return {
    type: 'transition-failed',
    timestamp: ts,
    transitionName: name,
    errorMessage: 'boom',
    exceptionType: 'RuntimeError',
  };
}

function transitionTimedOut(name: string, ts: number): NetEvent {
  return {
    type: 'transition-timed-out',
    timestamp: ts,
    transitionName: name,
    deadlineMs: 100,
    actualDurationMs: 150,
  };
}

function actionTimedOut(name: string, ts: number): NetEvent {
  return { type: 'action-timed-out', timestamp: ts, transitionName: name, timeoutMs: 100 };
}

function logMessage(level: string, ts: number): NetEvent {
  return {
    type: 'log-message',
    timestamp: ts,
    transitionName: 'Process',
    logger: 'test',
    level,
    message: `${level} log line`,
    error: null,
    errorMessage: null,
  };
}

describe('SessionArchive v2 (libpetri 1.7.0)', () => {
  const writer = new SessionArchiveWriter();
  const reader = new SessionArchiveReader();

  describe('default write format', () => {
    it('should default to v2', () => {
      const registry = new DebugSessionRegistry();
      registry.register('default-v2', TEST_NET, { channel: 'voice' });

      const compressed = writer.write(registry.getSession('default-v2')!);
      const metadata = reader.readMetadata(compressed);

      expect(metadata.version).toBe(2);
      expect(metadata.version).toBe(CURRENT_VERSION);
    });
  });

  describe('tags + endTime round-trip', () => {
    it('should preserve tags and endTime across round-trip', () => {
      const registry = new DebugSessionRegistry();
      const session = registry.register('voice-1', TEST_NET, {
        channel: 'voice',
        env: 'staging',
      });

      const t0 = Date.parse('2026-04-15T10:00:00Z');
      session.eventStore.append(transitionEnabled('Process', t0));
      session.eventStore.append(transitionStarted('Process', t0 + 10));
      session.eventStore.append(transitionCompleted('Process', t0 + 50));
      registry.complete('voice-1');

      const compressed = writer.write(registry.getSession('voice-1')!);
      const imported = reader.readFull(compressed);

      expect(imported.metadata.version).toBe(2);
      // Narrow the union via `version === 2` so TS knows v2-only fields exist.
      if (imported.metadata.version !== 2) throw new Error('expected v2');
      const v2 = imported.metadata;

      expect(v2.sessionId).toBe('voice-1');
      expect(v2.tags).toEqual({ channel: 'voice', env: 'staging' });
      expect(v2.endTime).toBeDefined();
      expect(Date.parse(v2.endTime!)).toBeGreaterThanOrEqual(t0);
    });

    it('should round-trip an empty session as v2 with empty tags', () => {
      const registry = new DebugSessionRegistry();
      registry.register('empty-v2', TEST_NET, { channel: 'text' });

      const compressed = writer.write(registry.getSession('empty-v2')!);
      const imported = reader.readFull(compressed);

      if (imported.metadata.version !== 2) throw new Error('expected v2');
      expect(imported.metadata.eventCount).toBe(0);
      expect(imported.metadata.tags).toEqual({ channel: 'text' });
      expect(imported.metadata.metadata.eventTypeHistogram).toEqual({});
      expect(imported.metadata.metadata.hasErrors).toBe(false);
      expect(imported.metadata.endTime).toBeUndefined();
    });
  });

  describe('pre-computed metadata histogram', () => {
    it('should compute histogram with correct counts per event type', () => {
      const registry = new DebugSessionRegistry();
      const session = registry.register('hist-1', TEST_NET);

      const t0 = Date.parse('2026-04-15T10:00:00Z');
      session.eventStore.append(transitionEnabled('Process', t0));
      session.eventStore.append(transitionStarted('Process', t0 + 10));
      session.eventStore.append(transitionStarted('Process', t0 + 20));
      session.eventStore.append(transitionCompleted('Process', t0 + 50));

      const compressed = writer.write(session);
      const imported = reader.readFull(compressed);

      if (imported.metadata.version !== 2) throw new Error('expected v2');
      const histogram = imported.metadata.metadata.eventTypeHistogram;

      expect(histogram['TransitionEnabled']).toBe(1);
      expect(histogram['TransitionStarted']).toBe(2);
      expect(histogram['TransitionCompleted']).toBe(1);
      // Absent types should not appear in the histogram
      expect(histogram['TokenAdded']).toBeUndefined();
    });

    it('should emit histogram keys in alphabetical order for determinism', () => {
      const registry = new DebugSessionRegistry();
      const session = registry.register('det-1', TEST_NET);

      const t0 = Date.parse('2026-04-15T10:00:00Z');
      // Append in a weird order to make sure sort happens.
      session.eventStore.append(transitionCompleted('Process', t0));
      session.eventStore.append(transitionEnabled('Process', t0 + 1));
      session.eventStore.append(transitionStarted('Process', t0 + 2));

      const metadata = computeMetadata(session.eventStore);
      const keys = Object.keys(metadata.eventTypeHistogram);
      const sorted = [...keys].sort();
      expect(keys).toEqual(sorted);
    });

    it('should track first and last event timestamps', () => {
      const registry = new DebugSessionRegistry();
      const session = registry.register('ts-1', TEST_NET);

      const t0 = Date.parse('2026-04-15T10:00:00Z');
      session.eventStore.append(transitionEnabled('Process', t0));
      session.eventStore.append(transitionStarted('Process', t0 + 100));
      session.eventStore.append(transitionCompleted('Process', t0 + 200));

      const compressed = writer.write(session);
      const imported = reader.readFull(compressed);

      if (imported.metadata.version !== 2) throw new Error('expected v2');
      expect(imported.metadata.metadata.firstEventTime).toBe(new Date(t0).toISOString());
      expect(imported.metadata.metadata.lastEventTime).toBe(new Date(t0 + 200).toISOString());
    });
  });

  describe('hasErrors detection', () => {
    it.each([
      ['TransitionFailed', transitionFailed],
      ['TransitionTimedOut', transitionTimedOut],
      ['ActionTimedOut', actionTimedOut],
    ] as const)('should flag hasErrors=true for %s', (_label, factory) => {
      const registry = new DebugSessionRegistry();
      const session = registry.register(`err-${_label}`, TEST_NET);
      const t0 = Date.parse('2026-04-15T10:00:00Z');
      session.eventStore.append(factory('Process', t0));

      const compressed = writer.write(session);
      const imported = reader.readFull(compressed);

      if (imported.metadata.version !== 2) throw new Error('expected v2');
      expect(imported.metadata.metadata.hasErrors).toBe(true);
    });

    it('should flag hasErrors=true for LogMessage at ERROR level', () => {
      const registry = new DebugSessionRegistry();
      const session = registry.register('err-log', TEST_NET);
      session.eventStore.append(logMessage('ERROR', Date.parse('2026-04-15T10:00:00Z')));

      const compressed = writer.write(session);
      const imported = reader.readFull(compressed);

      if (imported.metadata.version !== 2) throw new Error('expected v2');
      expect(imported.metadata.metadata.hasErrors).toBe(true);
    });

    it('should flag hasErrors=true for LogMessage at lowercase error level', () => {
      const registry = new DebugSessionRegistry();
      const session = registry.register('err-log-lower', TEST_NET);
      session.eventStore.append(logMessage('error', Date.parse('2026-04-15T10:00:00Z')));

      const compressed = writer.write(session);
      const imported = reader.readFull(compressed);

      if (imported.metadata.version !== 2) throw new Error('expected v2');
      expect(imported.metadata.metadata.hasErrors).toBe(true);
    });

    it('should leave hasErrors=false for non-error events and INFO logs', () => {
      const registry = new DebugSessionRegistry();
      const session = registry.register('clean', TEST_NET);
      const t0 = Date.parse('2026-04-15T10:00:00Z');
      session.eventStore.append(transitionEnabled('Process', t0));
      session.eventStore.append(logMessage('INFO', t0 + 1));
      session.eventStore.append(transitionCompleted('Process', t0 + 2));

      const compressed = writer.write(session);
      const imported = reader.readFull(compressed);

      if (imported.metadata.version !== 2) throw new Error('expected v2');
      expect(imported.metadata.metadata.hasErrors).toBe(false);
    });
  });

  describe('v1 backward compatibility', () => {
    it('should write v1 when writeV1() is called explicitly', () => {
      const registry = new DebugSessionRegistry();
      const session = registry.register('legacy', TEST_NET, { channel: 'voice' });
      session.eventStore.append(transitionEnabled('Process', Date.now()));

      const compressed = writer.writeV1(session);
      const imported = reader.readFull(compressed);

      expect(imported.metadata.version).toBe(1);
      expect(imported.metadata.sessionId).toBe('legacy');
      expect(imported.eventStore.eventCount()).toBe(1);
      // v1 has no tags / endTime / metadata — narrow via discriminant.
      if (imported.metadata.version === 2) throw new Error('expected v1');
      // Accessing v2-only fields on v1 would be a compile error thanks to narrowing.
    });

    it('should read mixed v1 and v2 archives from the same reader', () => {
      const registry = new DebugSessionRegistry();
      const v1Session = registry.register('v1-session', TEST_NET);
      v1Session.eventStore.append(transitionEnabled('Process', Date.now()));

      const v2Session = registry.register('v2-session', TEST_NET, { channel: 'voice' });
      v2Session.eventStore.append(transitionEnabled('Process', Date.now()));

      const v1Bytes = writer.writeV1(v1Session);
      const v2Bytes = writer.write(v2Session); // default v2

      const v1Read = reader.readFull(v1Bytes);
      const v2Read = reader.readFull(v2Bytes);

      expect(v1Read.metadata.version).toBe(1);
      expect(v2Read.metadata.version).toBe(2);
      expect(v1Read.metadata.sessionId).toBe('v1-session');
      expect(v2Read.metadata.sessionId).toBe('v2-session');

      if (v2Read.metadata.version === 2) {
        expect(v2Read.metadata.tags).toEqual({ channel: 'voice' });
      }
    });
  });

  describe('unsupported version rejection', () => {
    it('should reject a hand-crafted header with version=99', () => {
      // Build a minimal v2-shaped header but with version=99 to simulate an
      // archive produced by a future libpetri version the current reader
      // can't understand.
      const badHeader = {
        version: 99,
        sessionId: 'x',
        netName: 'n',
        dotDiagram: 'digraph{}',
        startTime: new Date().toISOString(),
        eventCount: 0,
        structure: { places: [], transitions: [] },
      };
      const headerJson = Buffer.from(JSON.stringify(badHeader), 'utf-8');
      const metaLen = Buffer.alloc(4);
      metaLen.writeUInt32BE(headerJson.length);
      const raw = Buffer.concat([metaLen, headerJson]);
      const compressed = gzipSync(raw);

      expect(() => reader.readMetadata(compressed)).toThrow(
        /Unsupported archive version: 99/,
      );
      expect(() => reader.readMetadata(compressed)).toThrow(
        new RegExp(`${MIN_SUPPORTED_VERSION}..${CURRENT_VERSION}`),
      );
    });
  });

  describe('on-wire shape', () => {
    it('should serialize tags and metadata as first-class header fields', () => {
      const registry = new DebugSessionRegistry();
      const session = registry.register('wire-1', TEST_NET, { channel: 'voice' });
      session.eventStore.append(transitionEnabled('Process', Date.parse('2026-04-15T10:00:00Z')));

      const compressed = writer.write(session);
      const raw = gunzipSync(compressed);
      const metaLen = raw.readUInt32BE(0);
      const headerJson = raw.subarray(4, 4 + metaLen).toString('utf-8');
      const headerObj = JSON.parse(headerJson) as SessionArchiveV2;

      expect(headerObj.version).toBe(2);
      expect(headerObj.tags).toEqual({ channel: 'voice' });
      expect(headerObj.metadata).toBeDefined();
      expect(headerObj.metadata.eventTypeHistogram['TransitionEnabled']).toBe(1);
      expect(headerObj.metadata.hasErrors).toBe(false);
    });
  });
});

File: java/src/test/java/org/libpetri/debug/archive/SessionArchiveV2Test.java
package org.libpetri.debug.archive;

import org.libpetri.core.Arc.In;
import org.libpetri.core.Arc.Out;
import org.libpetri.core.PetriNet;
import org.libpetri.core.Place;
import org.libpetri.core.Token;
import org.libpetri.core.Transition;
import org.libpetri.debug.DebugSessionRegistry;
import org.libpetri.event.NetEvent;
import org.junit.jupiter.api.Test;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.time.Duration;
import java.time.Instant;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Round-trip tests for the libpetri 1.7.0 v2 archive format. These focus on the
 * format-version-specific behavior (richer header, pre-computed metadata, mixed bucket
 * coexistence) — the v1 round-trip is exercised by {@link SessionArchiveRoundTripTest}.
 */
class SessionArchiveV2Test {

    private static final Place<String> INPUT = Place.of("Input", String.class);
    private static final Place<String> OUTPUT = Place.of("Output", String.class);

    private static final PetriNet TEST_NET = PetriNet.builder("TestNet")
            .transitions(
                    Transition.builder("Process")
                            .inputs(In.one(INPUT))
                            .outputs(Out.place(OUTPUT))
                            .build()
            )
            .build();

    // ======================== v2 round-trip ========================

    @Test
    void shouldDefaultToV2Format() throws IOException {
        var session = registerWithEvents("default-v2", Map.of("channel", "voice"));

        var archive = writeAndReadHeader(session);

        assertEquals(SessionArchive.CURRENT_VERSION, archive.version(), "default writer should emit current version");
        assertEquals(2, archive.version(), "current version should be 2 in libpetri 1.7.0");
        assertInstanceOf(SessionArchive.V2.class, archive,
            "default writer should produce a V2 record so callers can pattern-match on tags/endTime/metadata");
    }

    @Test
    void shouldRoundTripTagsAndEndTime() throws IOException {
        var registry = new DebugSessionRegistry();
        var session = registry.register("voice-1", TEST_NET, Map.of("channel", "voice", "env", "staging"));
        appendNonErrorEvents(session);
        registry.complete("voice-1");
        var completed = registry.getSession("voice-1").orElseThrow();

        var imported = writeAndReadFull(completed);

        // Pattern match to the v2 subtype to access v2-only fields with type safety.
        var v2 = assertInstanceOf(SessionArchive.V2.class, imported.metadata());
        assertEquals("voice-1", v2.sessionId());
        assertEquals(Map.of("channel", "voice", "env", "staging"), v2.tags());
        assertNotNull(v2.endTime(), "endTime should be preserved across the round-trip");
        assertNotNull(v2.durationMs(), "durationMs derived from start/end should be available");
        assertTrue(v2.durationMs() >= 0);
    }

    @Test
    void shouldRoundTripPreComputedMetadata() throws IOException {
        var registry = new DebugSessionRegistry();
        var session = registry.register("metadata-1", TEST_NET);

        // Mix of event types so the histogram has multiple buckets.
        var t0 = Instant.parse("2026-04-15T10:00:00Z");
        session.eventStore().append(new NetEvent.ExecutionStarted(t0, "TestNet", "exec-1"));
        session.eventStore().append(new NetEvent.TransitionEnabled(t0.plusMillis(10), "Process"));
        session.eventStore().append(new NetEvent.TransitionStarted(t0.plusMillis(20), "Process",
                List.of(Token.of("hello"))));
        session.eventStore().append(new NetEvent.TransitionStarted(t0.plusMillis(30), "Process",
                List.of(Token.of("world"))));
        session.eventStore().append(new NetEvent.TransitionCompleted(t0.plusMillis(50), "Process",
                List.of(Token.of("done")), Duration.ofMillis(20)));

        var imported = writeAndReadFull(session);
        var v2 = assertInstanceOf(SessionArchive.V2.class, imported.metadata());
        var metadata = v2.metadata();

        assertNotNull(metadata, "v2 archive must carry pre-computed metadata");
        assertEquals(1L, metadata.eventTypeHistogram().get("ExecutionStarted"));
        assertEquals(1L, metadata.eventTypeHistogram().get("TransitionEnabled"));
        assertEquals(2L, metadata.eventTypeHistogram().get("TransitionStarted"));
        assertEquals(1L, metadata.eventTypeHistogram().get("TransitionCompleted"));
        assertNull(metadata.eventTypeHistogram().get("TokenAdded"), "absent types should not appear");
        assertEquals(t0, metadata.firstEventTime());
        assertEquals(t0.plusMillis(50), metadata.lastEventTime());
        assertFalse(metadata.hasErrors(), "no error events were appended");
    }

    @Test
    void shouldDetectErrorsInMetadata() throws IOException {
        var registry = new DebugSessionRegistry();
        var session = registry.register("err-1", TEST_NET);
        var t0 = Instant.parse("2026-04-15T10:00:00Z");

        session.eventStore().append(new NetEvent.ExecutionStarted(t0, "TestNet", "exec-1"));
        session.eventStore().append(new NetEvent.TransitionFailed(
                t0.plusMillis(10), "Process", "boom", "java.lang.IllegalStateException"));

        var imported = writeAndReadFull(session);
        var v2 = assertInstanceOf(SessionArchive.V2.class, imported.metadata());
        assertTrue(v2.metadata().hasErrors(), "TransitionFailed must flag hasErrors");
    }

    @Test
    void shouldDetectErrorLogLevelInMetadata() throws IOException {
        var registry = new DebugSessionRegistry();
        var session = registry.register("err-log", TEST_NET);
        var t0 = Instant.parse("2026-04-15T10:00:00Z");
        session.eventStore().append(new NetEvent.LogMessage(
                t0, "Process", "test.Logger", "ERROR", "kaboom", null, null));

        var imported = writeAndReadFull(session);
        var v2 = assertInstanceOf(SessionArchive.V2.class, imported.metadata());
        assertTrue(v2.metadata().hasErrors(), "LogMessage at ERROR level must flag hasErrors");
    }

    @Test
    void shouldRoundTripEmptySessionAsV2() throws IOException {
        var registry = new DebugSessionRegistry();
        var session = registry.register("empty-v2", TEST_NET, Map.of("channel", "text"));

        var imported = writeAndReadFull(session);
        var v2 = assertInstanceOf(SessionArchive.V2.class, imported.metadata());

        assertEquals(0, v2.eventCount());
        assertEquals(0, imported.eventStore().eventCount());
        assertEquals(Map.of("channel", "text"), v2.tags());
        assertNull(v2.endTime(), "active session must have no endTime");
        assertNull(v2.durationMs(), "active session must have no duration");
        assertNotNull(v2.metadata(), "metadata should never be null even for empty sessions");
        assertTrue(v2.metadata().eventTypeHistogram().isEmpty());
        assertNull(v2.metadata().firstEventTime());
    }

    // ======================== mixed v1/v2 reading ========================

    @Test
    void shouldReadV1ArchiveAfterRefactor() throws IOException {
        var registry = new DebugSessionRegistry();
        var session = registry.register("legacy-v1", TEST_NET);
        session.eventStore().append(new NetEvent.TransitionEnabled(Instant.now(), "Process"));

        var writer = new SessionArchiveWriter();
        var buf = new ByteArrayOutputStream();
        writer.writeV1(session, buf);

        var imported = new SessionArchiveReader().readFull(new ByteArrayInputStream(buf.toByteArray()));

        var v1 = assertInstanceOf(SessionArchive.V1.class, imported.metadata());
        assertEquals(1, v1.version());
        assertEquals("legacy-v1", v1.sessionId());
        assertEquals(1, imported.eventStore().eventCount());

        // v2-only accessors return defaults for v1 archives via the sealed interface.
        assertEquals(Map.of(), v1.tags());
        assertNull(v1.endTime());
        assertNull(v1.metadata(), "v1 has no pre-computed metadata; callers should use SessionMetadata.fromEvents()");
    }

    @Test
    void shouldReadMixedV1AndV2InSameBucket() throws IOException {
        var registry = new DebugSessionRegistry();

        var v1Session = registry.register("legacy", TEST_NET);
        v1Session.eventStore().append(new NetEvent.TransitionEnabled(Instant.now(), "Process"));
        var v1Bytes = new ByteArrayOutputStream();
        new SessionArchiveWriter().writeV1(v1Session, v1Bytes);

        var v2Session = registry.register("modern", TEST_NET, Map.of("channel", "voice"));
        v2Session.eventStore().append(new NetEvent.TransitionEnabled(Instant.now(), "Process"));
        var v2Bytes = new ByteArrayOutputStream();
        new SessionArchiveWriter().write(v2Session, v2Bytes);

        var reader = new SessionArchiveReader();
        var v1Read = reader.readFull(new ByteArrayInputStream(v1Bytes.toByteArray()));
        var v2Read = reader.readFull(new ByteArrayInputStream(v2Bytes.toByteArray()));

        assertInstanceOf(SessionArchive.V1.class, v1Read.metadata());
        assertInstanceOf(SessionArchive.V2.class, v2Read.metadata());
        assertEquals("legacy", v1Read.metadata().sessionId());
        assertEquals("modern", v2Read.metadata().sessionId());
        assertEquals(Map.of("channel", "voice"), v2Read.metadata().tags());
        assertEquals(Map.of(), v1Read.metadata().tags(), "v1 default tag accessor must be empty");
    }

    @Test
    void shouldRejectUnsupportedVersion() throws IOException {
        // Hand-craft a header with version=99 by abusing the V2 record's lenient deserializer.
        // We can't use writeV1/write because both pin the version. Instead, write a header
        // by directly encoding bytes. Easier: construct a malformed V2 instance and check
        // that the deserializer rejects unknown versions on read.
        var v2Json = "{\"version\":99,\"sessionId\":\"x\",\"netName\":\"n\",\"dotDiagram\":\"\","
                + "\"startTime\":\"2026-04-15T10:00:00Z\",\"eventCount\":0,"
                + "\"structure\":{\"places\":[],\"transitions\":[]}}";
        var headerBytes = v2Json.getBytes(java.nio.charset.StandardCharsets.UTF_8);

        var raw = new ByteArrayOutputStream();
        try (var dataOut = new java.io.DataOutputStream(
                new net.jpountz.lz4.LZ4FrameOutputStream(raw))) {
            dataOut.writeInt(headerBytes.length);
            dataOut.write(headerBytes);
        }

        var reader = new SessionArchiveReader();
        var ex = assertThrows(IOException.class,
            () -> reader.readMetadata(new ByteArrayInputStream(raw.toByteArray())));
        assertTrue(ex.getMessage().contains("Unsupported archive version"),
            "expected unsupported-version error, got: " + ex.getMessage());
        assertTrue(ex.getMessage().contains("99"));
    }

    // ======================== SessionMetadata.fromEvents (v1 fallback) ========================

    @Test
    void sessionMetadataFromEventsShouldComputeOnDemandForV1() {
        var registry = new DebugSessionRegistry();
        var session = registry.register("v1-fallback", TEST_NET);
        var t0 = Instant.parse("2026-04-15T10:00:00Z");
        session.eventStore().append(new NetEvent.ExecutionStarted(t0, "TestNet", "exec-1"));
        session.eventStore().append(new NetEvent.TransitionFailed(
                t0.plusMillis(20), "Process", "x", "T"));

        var metadata = SessionMetadata.fromEvents(session.eventStore().events());

        assertEquals(t0, metadata.firstEventTime());
        assertEquals(t0.plusMillis(20), metadata.lastEventTime());
        assertTrue(metadata.hasErrors());
        assertEquals(1L, metadata.eventTypeHistogram().get("ExecutionStarted"));
        assertEquals(1L, metadata.eventTypeHistogram().get("TransitionFailed"));
    }

    // ======================== helpers ========================

    private DebugSessionRegistry.DebugSession registerWithEvents(String id, Map<String, String> tags) {
        var registry = new DebugSessionRegistry();
        var session = registry.register(id, TEST_NET, tags);
        appendNonErrorEvents(session);
        return session;
    }

    private void appendNonErrorEvents(DebugSessionRegistry.DebugSession session) {
        var t0 = Instant.parse("2026-04-15T10:00:00Z");
        session.eventStore().append(new NetEvent.ExecutionStarted(t0, "TestNet", "exec-1"));
        session.eventStore().append(new NetEvent.TransitionStarted(t0.plusMillis(10), "Process",
                List.of(Token.of("hello"))));
        session.eventStore().append(new NetEvent.TransitionCompleted(t0.plusMillis(50), "Process",
                List.of(Token.of("world")), Duration.ofMillis(40)));
    }

    private SessionArchive writeAndReadHeader(DebugSessionRegistry.DebugSession session) throws IOException {
        var buf = new ByteArrayOutputStream();
        new SessionArchiveWriter().write(session, buf);
        return new SessionArchiveReader().readMetadata(new ByteArrayInputStream(buf.toByteArray()));
    }

    private SessionArchiveReader.ImportedSession writeAndReadFull(DebugSessionRegistry.DebugSession session) throws IOException {
        var buf = new ByteArrayOutputStream();
        new SessionArchiveWriter().write(session, buf);
        return new SessionArchiveReader().readFull(new ByteArrayInputStream(buf.toByteArray()));
    }
}
